src/lakefs_spec/spec.py

```diff
@@ -329,6 +384,9 @@ def put_file(
     ):
         repository, branch, resource = parse(rpath)
 
+        if self.create_branch_ok:
+            ensure_branch(self.client, repository, branch, self.source_branch)
+
```

Comment:
This will probably result in endpoint thrashing if you upload 2000 data frame shards.

Should we agree on only ensuring the branch once in `fs.put()`? After all, `put_file` is more of a lower-level method, that should probably not be called directly.